orderbook/src/bad_token.rs

```diff
@@ -0,0 +1,465 @@
+                    reason: "decimals too large".to_string(),
+                });
+            }
+        };
+
+        let take_from = match self.find_address_owning_token(token).await? {
+            Some((address, balance)) if balance > amount => {
+                tracing::debug!("testing token {:?} with pool {:?}", token, address);
+                address
+            }
+            _ => {
+                return Ok(TokenInfo::Bad {
+                    reason: "no amm pool with base tokens with enough balance".to_string(),
+                });
+            }
+        };
+        let request = self.create_trace_request(token, amount, take_from);
+        let traces = trace_many::trace_many(request, &self.web3).await?;
+        self.handle_response(&traces, amount)
+    }
+
+    // Based on amm pools find the address with the largest amount of the token.
+    // Err if communication with the node failed.
+    // Ok(None) if there is no pool or getting the balance fails.
+    // Ok(address, balance) for an address that has this amount of balance of the token.
+    async fn find_address_owning_token(&self, token: H160) -> Result<Option<(H160, U256)>> {
+        const BATCH_SIZE: usize = 100;
+
+        let pairs = self
+            .base_tokens
+            .iter()
+            .filter_map(move |&base_token| TokenPair::new(base_token, token));
+        let candidates = pairs
+            .flat_map(|pair| self.pools.iter().map(move |pool| pool.pair_address(&pair)))
+            .collect::<HashSet<_>>();
+
+        let instance = ERC20::at(&self.web3, token);
+        let mut batch = CallBatch::new(self.web3.transport());
+        let futures = candidates
+            .iter()
+            .map(|&address| {
+                let fut = instance.balance_of(address).batch_call(&mut batch);
+                async move { (address, fut.await) }
+            })
+            .collect::<Vec<_>>();
+        batch.execute_all(BATCH_SIZE).await;
+
+        let mut biggest_balance = None;
+        for future in futures {
+            let (address, result) = future.await;
+            let balance = match result {
+                Ok(balance) => balance,
+                Err(err) => {
+                    return match EthcontractError::classify(&err) {
+                        EthcontractError::Node => Err(err.into()),
+                        EthcontractError::Contract => Ok(None),
+                    }
+                }
+            };
+            match biggest_balance {
+                Some((_, current_biggest)) if current_biggest > balance => (),
+                _ => biggest_balance = Some((address, balance)),
+            }
+        }
+        Ok(biggest_balance)
+    }
+
+    fn create_trace_request(&self, token: H160, amount: U256, take_from: H160) -> Vec<CallRequest> {
+        let instance = ERC20::at(&self.web3, token);
+
+        // We transfer one unit (base on decimals) of the token from the amm pool into the
```

Comment:
Is this true? Don't we transfer `amount` of token?

Reply:
From the perspective of this function it is amount but where it is called it is always set to this. Will change it to say amount here.